src/server/engine_factory.rs

```diff
@@ -192,8 +192,8 @@ impl TabletFactory<RocksEngine> for KvEngineFactory {
         let tablet_name = path.file_name().unwrap().to_str().unwrap().to_string();
         db_opts.set_info_log(TabletLogger::new(tablet_name));
         let cf_opts = self.cf_opts(EngineType::RaftKv2);
-        if let Some(listener) = &self.inner.flow_listener && let Some(suffix) = ctx.suffix {
-            db_opts.add_event_listener(listener.clone_with(ctx.id, suffix));
+        if let Some(listener) = &self.inner.flow_listener {
```

Comment:
Note this function can be called whenever necessary. So it might not be necessary to open a true tablet. Any operations depend on the event while is not guarded by ref count will probably go wrong.

Reply:
Got it. The flow control does not rely on the Tablet object anymore by your improvement.